kore/src/Kore/Step/Rule.hs

```diff
@@ -916,3 +922,45 @@ applySubstitution substitution rule =
     subst = Substitution.toMap substitution
     finalRule = substitute subst rule
     substitutedVariables = Substitution.variables substitution
+
+-- | The typeclasses 'ToRulePattern' and 'FromRulePattern' are intended to
+-- be implemented by types which contain more (or the same amount of)
+-- information as 'RulePattern Variable'.
+class ToRulePattern rule where
+    toRulePattern :: rule -> RulePattern Variable
+    default toRulePattern
+        :: Coercible rule (RulePattern Variable)
+        => rule -> RulePattern Variable
+    toRulePattern = coerce
+
+instance ToRulePattern (OnePathRule Variable)
+
+instance ToRulePattern (AllPathRule Variable)
+
+instance ToRulePattern (ImplicationRule Variable)
+
+instance ToRulePattern (ReachabilityRule Variable) where
+    toRulePattern (OnePath rule) = toRulePattern rule
+    toRulePattern (AllPath rule) = toRulePattern rule
+
+-- | We need to know the context when transforming a 'RulePattern Variable'
+-- into a 'rule', hence the first 'rule' argument. In general it can be ignored
+-- when the 'rule' and the 'RulePattern Variable' are representationally equal.
```

Comment:
Can the first argument be a `Proxy` then?

Reply:
I don't think it can, or I just don't know how to do this with a `Proxy`. I need a case on that first argument, if it's `OnePathRule` apply the `OnePath` constructor and if it's an `AllPathRule` apply the `AllPath` constructor.